### PR TITLE
Move C# tests to testing metadata emitted attributes through `CompileAndVerify`

### PR DESCRIPTION
**Pull request by OmarTawfik**
Contributes to #18799

This PR has the C# part of the fixes:

* Removed the unnecessary/wrong helpers, and replaced all with `CompileAndVerify()`.
* Fixed a bug in `CompileAndVerify()` where it was not loading all modules of a multi-module compilation during symbol verification.

Please note that it is quite possible that some tests still exist testing source attributes directly, however, that was the intended behavior, and a manual pass might be required to identify, and possibly change them.

### CODE REVIEW

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-compiler for review.

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-compiler for review.

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-compiler for compiler test changes.

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-compiler can I get review here?
VB changes should follow in a separate PR.

**Comment by AlekseyTs** [comment]
Are there VB tests with the same problem?

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests.cs [hunk 1]
```diff
@@ -1189,13 +1186,10 @@ static void Main()
...
                 // verify no ParamArrayAttribute on p2
-                if (isFromSource)
-                {
-                    WellKnownAttributesTestBase.VerifyParamArrayAttribute(parameters[1], (SourceModuleSymbol)moduleSymbol, expected: false);
-                }
+                VerifyParamArrayAttribute(parameters[1], expected: false);
             };
 
-            CompileAndVerify(source, sourceSymbolValidator: symbolValidator(true), symbolValidator: symbolValidator(false));
+            CompileAndVerify(source, symbolValidator: symbolValidator);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests.cs]: >CompileAndVerify(source, symbolValidator: symbolValidator); [](start = 12, length = 59)

This change removes testing for source, I think we should put it back.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests.cs]: I mean we should still pass sourceSymbolValidator.

---
In reply to: [150616316](https://github.com/dotnet/roslyn/pull/22896#discussion_r150616316) [](ancestors = 150616316)

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Dynamic.cs [hunk 2]
```diff
@@ -120,37 +116,32 @@ internal struct DynamicAttributeValidator
             private readonly NamedTypeSymbol _synthesizedMyDelegateType;
             private bool[] _expectedTransformFlags;
 
-            private DynamicAttributeValidator(CSharpCompilation compilation)
+            private DynamicAttributeValidator(ModuleSymbol module)
             {
-                _comp = compilation;
-                _srcAssembly = compilation.SourceAssembly;
-                NamespaceSymbol globalNamespace = _srcAssembly.Modules[0].GlobalNamespace;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Dynamic.cs]: >globalNamespace [](start = 32, length = 15)

It feels like we could keep this local (with adjusted initializer) and avoid many unnecessary changes below.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Security.cs [hunk 3]
```diff
@@ -896,6 +900,7 @@ public class C
                         "\u0005" + // length of UTF-8 string (small enough to fit in 1 byte)
                         "User2", // argument value (@"User2")
                 });
+        });
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Security.cs]: The offset looks off.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Security.cs [hunk 4]
```diff
@@ -1284,84 +1292,95 @@ public class C
...
-            {
-                var assembly = module.ContainingAssembly;
-                var ns = module.GlobalNamespace.GetMember<NamespaceSymbol>("N");
-                var namedType = ns.GetMember<NamedTypeSymbol>("C");
-                var type = (Cci.ITypeDefinition)namedType;
-                var method = (Cci.IMethodDefinition)namedType.GetMember("Goo");
 
-                if (isFromSource)
+            var compilation = CreateStandardCompilation(source, options: TestOptions.UnsafeReleaseDll);
+            CompileAndVerify(compilation, symbolValidator: module =>
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Security.cs]: >CompileAndVerify [](start = 12, length = 16)

We should keep source validation and pass sourceSymbolValidator

**Comment by AlekseyTs** [comment]
>     /// TODO: Verification was disabled for net modules for now. Add it back once module support has been added.

Is there an issue tracking this work?

---
Refers to: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Synthesized.cs:24 in 117522f. [](commit_id = 117522f908d58b3b248efae05bec847f2b6c8cdf, deletion_comment = False)

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Tuples.cs [hunk 5]
```diff
@@ -284,30 +292,25 @@ private static string ToTestString(Symbol symbol)
...
+                _derivedClass;
+
+            private TupleAttributeValidator(ModuleSymbol module)
             {
-                _tupleAttrTransformNames = (MethodSymbol)compilation.GetWellKnownTypeMember(
-                    WellKnownMember.System_Runtime_CompilerServices_TupleElementNamesAttribute__ctorTransformNames);
-                Assert.NotNull(_tupleAttrTransformNames);
-
-                _comp = compilation;
-                var globalNs = compilation.SourceModule.GlobalNamespace;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Tuples.cs]: >globalNs [](start = 20, length = 8)

It feels like we could keep this local.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs [hunk 6]
```diff
@@ -4742,29 +4742,26 @@ public static void Main()
...
-
-                var noParamsMethod = (MethodSymbol)type.GetMember("NoParams");
-                var noParamsParameter = noParamsMethod.Parameters[0];
-                Assert.Equal(0, noParamsParameter.GetSynthesizedAttributes().Length);
-            };
-
             // Verify attributes from source and then load metadata to see attributes are written correctly.
             var comp = CompileAndVerify(
                 compilation,
-                sourceSymbolValidator: attributeValidator,
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs]: >sourceSymbolValidator: attributeValidator, [](start = 16, length = 42)

It feels like we should keep source validator.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOverridingAndHiding.cs [hunk 7]
```diff
@@ -3984,14 +3981,10 @@ static void Main()
...
 
-                if (isFromSource)
-                {
-                    var srcModule = (SourceModuleSymbol)module;
-                    WellKnownAttributesTestBase.VerifyParamArrayAttribute(parameterB, srcModule);
-                }
+                WellKnownAttributesTestBase.VerifyParamArrayAttribute(parameterB);
             };
 
-            var verifier = CompileAndVerify(source, symbolValidator: validator(false), sourceSymbolValidator: validator(true), expectedOutput: @"System.Int32[]");
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOverridingAndHiding.cs]: >sourceSymbolValidator [](start = 87, length = 21)

We should keep sourceSymbolValidator.

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs [hunk 8]
```diff
@@ -4984,5 +5000,28 @@ public override void Method(T input)
                 //     public override void Method(T input)
                 Diagnostic(ErrorCode.ERR_CantChangeReturnTypeOnOverride, "Method").WithArguments("DerivingClass<T>.Method(T)", "BaseClass<T>.Method(T)", "int").WithLocation(11, 26));
         }
+
+        [Fact]
+        public void CompileAndVerifyModuleInlucesAllModules()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs]: >Inluces [](start = 42, length = 7)

Typo "Inluces"?

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs [hunk 9]
```diff
@@ -4984,5 +5000,28 @@ public override void Method(T input)
...
+        {
+            // Before this change, CompileAndVerify() didn't include other modules when testing a PEModule.
+            // Verify that symbols from other modules are accessible as well.
+
+            var modRef = ModuleMetadata.CreateFromImage(CreateStandardCompilation("public class A { }", options: TestOptions.ReleaseModule, assemblyName: "refMod").EmitToArray());
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs]: >EmitToArray [](start = 164, length = 11)

I think we have a helper EmitToImageReference or something similar that gives a ready to use reference.

File: src/Compilers/CSharp/Test/Emit/Emit/EmitMetadataTests.cs [hunk 10]
```diff
@@ -1233,49 +1233,42 @@ class C : B<string>
...
                 var classC = module.GlobalNamespace.GetTypeMembers("C").Single();
-                p = classC.BaseType.GetMembers("P").OfType<PropertySymbol>().Single();
-                VerifyAutoProperty(p, isFromSource);
+                p = classC.BaseType.GetProperty("P");
+                VerifyAutoProperty(p);
                 Assert.Equal(p.Type.SpecialType, SpecialType.System_String);
                 Assert.Equal(p.GetMethod.AssociatedSymbol, p);
             };
 
-            CompileAndVerify(source, sourceSymbolValidator: validator(true), symbolValidator: validator(false), options: TestOptions.ReleaseDll.WithMetadataImportOptions(MetadataImportOptions.Internal));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/EmitMetadataTests.cs]: >sourceSymbolValidator [](start = 37, length = 21)

It feels like we should keep sourceSymbolValidator

File: src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadingAttributes.cs [hunk 11]
```diff
@@ -1388,69 +1417,35 @@ .maxstack  8
...
-            var builder = GetDefaultPEBuilder(c1);
-
-            Assert.Empty(d1.GetAttributes());
-            Assert.Equal("System.Runtime.CompilerServices.DecimalConstantAttribute(0, 128, 0, 0, 7)", d1.GetCustomAttributesToEmit(builder).Single().ToString());
-            Assert.Equal(d1.ConstantValue, -7m);
-            Assert.Empty(m1.Parameters[0].GetAttributes());
-            Assert.Equal("System.Runtime.CompilerServices.DecimalConstantAttribute(0, 128, 0, 0, 7)", m1.Parameters[0].GetCustomAttributesToEmit(builder).Single().ToString());
-            Assert.Equal(m1.Parameters[0].ExplicitDefaultValue, -7m);
-
-            var c2 = CreateCompilationWithCustomILSource("", ilSource);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadingAttributes.cs]: >var c2 = CreateCompilationWithCustomILSource("", ilSource); [](start = 12, length = 59)

I think we should keep existing validation for symbols from c2 and c3. They are already PE symbols and it looks like the test asserts that API behavior is not order dependent.

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTestBase.cs [hunk 12]
```diff
@@ -385,22 +385,41 @@ internal static Alias Alias(DkmClrAliasKind kind, string name, string fullName,
...
-        }
+            var module = AssemblyMetadata.CreateFromImage(assembly).GetModules().Single().Module;
 
-        internal static SynthesizedAttributeData GetTupleElementNamesAttributeIfAny(IMethodSymbol method)
-        {
-            return GetAttributeIfAny(method, "System.Runtime.CompilerServices.TupleElementNamesAttribute");
-        }
+            var typeHandle = module
+                .GroupTypesByNamespaceOrThrow(StringComparer.Ordinal)
+                .SelectMany(group => group)
```
Comment (AlekseyTs) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTestBase.cs]: >.SelectMany(group => group) [](start = 16, length = 27)

This feels too complicated. Consider simply using module.MetadataReader.TypeDefinitions collection dirtectly.

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTestBase.cs [hunk 13]
```diff
@@ -385,22 +385,41 @@ internal static Alias Alias(DkmClrAliasKind kind, string name, string fullName,
...
+                .Single(handle => module.GetTypeDefNameOrThrow(handle) == method.ContainingType.Name);
```
Comment (AlekseyTs) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTestBase.cs]: >method.ContainingType.Name [](start = 74, length = 26)

Consider caching the name in a local.

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTestBase.cs [hunk 14]
```diff
@@ -385,22 +385,41 @@ internal static Alias Alias(DkmClrAliasKind kind, string name, string fullName,
...
 
-        internal static SynthesizedAttributeData GetAttributeIfAny(IMethodSymbol method, string typeName)
-        {
-            return ((MethodSymbol)method).GetSynthesizedAttributes(forReturnType: true).
-                Where(a => a.AttributeClass.ToTestDisplayString() == typeName).
-                SingleOrDefault();
+            var methodHandle = module
+                .GetMethodsOfTypeOrThrow(typeHandle)
+                .Single(handle => module.GetMethodDefNameOrThrow(handle) == method.Name);
```
Comment (AlekseyTs) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTestBase.cs]: >method.Name [](start = 76, length = 11)

Consider caching the name in a local.

File: src/Test/Utilities/Portable/CompilationVerifier.cs [hunk 15]
```diff
@@ -220,42 +233,24 @@ public CompilationVerifier VerifyDiagnostics(params DiagnosticDescription[] expe
...
-            return assemblies.Last();
-        }
-
-        private IEnumerable<IModuleSymbol> GetReferencesToModuleSymbols(IEnumerable<MetadataReference> references, MetadataImportOptions importOptions)
+        internal IModuleSymbol GetSymbolFromMetadata(MetadataReference metadataReference, MetadataImportOptions importOptions)
         {
             var dummy = _compilation
                 .RemoveAllSyntaxTrees()
-                .WithReferences(references)
+                .WithReferences(_compilation.References.Concat(new[] { metadataReference }))
```
Comment (AlekseyTs) [inline src/Test/Utilities/Portable/CompilationVerifier.cs]: >WithReferences [](start = 17, length = 14)

Should we use AddReference instead?

File: src/Test/Utilities/Portable/CommonTestBase.cs [hunk 16]
```diff
@@ -189,17 +189,22 @@ static internal void RunValidators(CompilationVerifier verifier, Action<PEAssemb
...
+                var assembly = ((AssemblyMetadata)emittedMetadata).GetAssembly();
+                assemblyValidator(assembly);
             }
 
             if (symbolValidator != null)
             {
-                var peModuleSymbol = verifier.GetModuleSymbolForEmittedImage();
-                Debug.Assert(peModuleSymbol != null);
-                symbolValidator(peModuleSymbol);
+                var emittedMetadata = verifier.GetMetadata();
```
Comment (AlekseyTs) [inline src/Test/Utilities/Portable/CommonTestBase.cs]: >verifier.GetMetadata() [](start = 38, length = 22)

Would it make sense to reuse result of calling this method across both if statements?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 9)

**Comment by OmarTawfik** [comment]
Yes. Almost the entire infrastructure is duplicated.

---
In reply to: [344001990](https://github.com/dotnet/roslyn/pull/22896#issuecomment-344001990) [](ancestors = 344001990)

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Security.cs (hunk 4 above)
Comment (OmarTawfik) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Security.cs]: For this perticular test, `SourceAssemblySymbol.GetSecurityAttributes` and `NamedTypeSymbol.GetSecurityInformation` are the exact kind of internal APIs that we are trying to get away from using in tests. Are you proposing testing security attributes for source symbols in a different way?

---
In reply to: [150625800](https://github.com/dotnet/roslyn/pull/22896#discussion_r150625800) [](ancestors = 150625800)

File: src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadingAttributes.cs (hunk 11 above)
Comment (OmarTawfik) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadingAttributes.cs]: @AlekseyTs I see in the IL above that these members has duplicate `DecimalConstantAttribute` attributes. (0, 128, 0, 0, 7) and (0, 128, 0, 0, 8). Is this intended? I don't see an explanation in the bug linked in the test.

---
In reply to: [150639861](https://github.com/dotnet/roslyn/pull/22896#discussion_r150639861) [](ancestors = 150639861)

**Comment by OmarTawfik** [comment]
>     /// TODO: Verification was disabled for net modules for now. Add it back once module support has been added.

No. I can log an issue, although I doubt there are any existing (or future) plans to modify peverify to support net modules. Maybe this can be true on day for netcore.
This is an existing comment moved from line 828. Will add a link to the issue in the comment.

---
In reply to: [344014521](https://github.com/dotnet/roslyn/pull/22896#issuecomment-344014521) [](ancestors = 344014521)

---
Refers to: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Synthesized.cs:24 in 117522f. [](commit_id = 117522f908d58b3b248efae05bec847f2b6c8cdf, deletion_comment = False)

**Comment by OmarTawfik** [comment]
@AlekseyTs  comments addressed.
@dotnet/roslyn-compiler  can I get a second review?

**Comment by jaredpar** [comment]
>         public static IEnumerable<object[]> OptimizationLevelTheoryData

Does xunit require that theory data be untyped like this? My intuition is that it would work with enum arrays here. 

---
Refers to: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Synthesized.cs:31 in 2800c0e. [](commit_id = 2800c0e2e617b5dcba2fda09a203dc886f625017, deletion_comment = False)

**Comment by OmarTawfik** [comment]
>         public static IEnumerable<object[]> OptimizationLevelTheoryData

Yes. XUnit analyzers dictate so (error xUnit1019: MemberData must reference a data type assignable to 'System.Collections.Generic.IEnumerable<object[]>'. The referenced type 'System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.OptimizationLevel[]>' is not valid.)

---
In reply to: [348362821](https://github.com/dotnet/roslyn/pull/22896#issuecomment-348362821) [](ancestors = 348362821)

---
Refers to: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Synthesized.cs:31 in 2800c0e. [](commit_id = 2800c0e2e617b5dcba2fda09a203dc886f625017, deletion_comment = False)

**Comment by OmarTawfik** [comment]
@jaredpar @AlekseyTs comments addressed.

**Comment by OmarTawfik** [comment]
@AlekseyTs any other comments?

File: src/Test/Utilities/Portable/CommonTestBase.cs [hunk 17]
```diff
@@ -194,19 +194,24 @@ internal CompilationVerifier CompileAndVerifyFieldMarshal(string source, Func<st
 
         static internal void RunValidators(CompilationVerifier verifier, Action<PEAssembly> assemblyValidator, Action<IModuleSymbol> symbolValidator)
         {
+            var emittedMetadata = verifier.GetMetadata();
```
Comment (AlekseyTs) [inline src/Test/Utilities/Portable/CommonTestBase.cs]: >var emittedMetadata = verifier.GetMetadata(); [](start = 12, length = 45)

We probably shouldn't even do this if there are no validators to run.

Comment (AlekseyTs) [reply src/Test/Utilities/Portable/CommonTestBase.cs]: I see that there is only one call site and it was adjusted to not call this method if there no validators. Consider adding an assert here.

---
In reply to: [155404430](https://github.com/dotnet/roslyn/pull/22896#discussion_r155404430) [](ancestors = 155404430)
